micrometer-spring-legacy/src/main/java/io/micrometer/spring/advice/TimedMethodInterceptor.java

```diff
@@ -0,0 +1,87 @@
+import io.micrometer.core.aop.TimedAspect;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.Timer.Builder;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import org.springframework.core.annotation.AnnotatedElementUtils;
+import org.springframework.util.ReflectionUtils;
+
+import java.lang.reflect.Method;
+
+public class TimedMethodInterceptor implements MethodInterceptor {
+    private final MeterRegistry registry;
+    private final TimedMetricNameResolver timedMetricNameResolver;
+    private final TimedTagsResolver timedTagsResolver;
+
+    public TimedMethodInterceptor(MeterRegistry registry) {
+        this(registry, invocation ->
+            Tags.of("class", invocation.getMethod().getDeclaringClass().getSimpleName(),
+                    "method", invocation.getMethod().getName())
+        );
+    }
+
+    public TimedMethodInterceptor(MeterRegistry registry, TimedTagsResolver timedTagsResolver) {
+        this(registry, (metricName, invocation) -> metricName, timedTagsResolver);
+    }
+
+    public TimedMethodInterceptor(MeterRegistry registry, TimedMetricNameResolver timedMetricNameResolver, TimedTagsResolver timedTagsResolver) {
+        this.registry = registry;
+        this.timedMetricNameResolver = timedMetricNameResolver;
+        this.timedTagsResolver = timedTagsResolver;
+    }
+
+    @Override
+    public Object invoke(MethodInvocation invocation) throws Throwable {
+        Method method = invocation.getMethod();
+        Timed methodLevelTimed = AnnotatedElementUtils.findMergedAnnotation(method, Timed.class);
+        Timed classLevelTimed = AnnotatedElementUtils.findMergedAnnotation(method.getDeclaringClass(), Timed.class);
+        if (methodLevelTimed == null) {
+            method = ReflectionUtils.findMethod(invocation.getThis().getClass(), method.getName(), method.getParameterTypes());
+            methodLevelTimed = AnnotatedElementUtils.findMergedAnnotation(method, Timed.class);
+        }
+        if (classLevelTimed == null && methodLevelTimed == null) {
+            return invocation.proceed();
+        }
+
+        Timer.Sample sample = Timer.start(registry);
+        final String metricName;
+        if (methodLevelTimed != null && !methodLevelTimed.value().isEmpty()) {
+            metricName = methodLevelTimed.value();
+        }
+        else if (classLevelTimed != null && !classLevelTimed.value().isEmpty()){
+            metricName = classLevelTimed.value();
+        }
+        else {
+            metricName = TimedAspect.DEFAULT_METRIC_NAME;
+        }
+        Builder builder = Timer.builder(timedMetricNameResolver.apply(metricName, invocation));
+        if (methodLevelTimed != null) {
+            builder.description(methodLevelTimed.description().isEmpty() ? null : methodLevelTimed.description())
```

Comment:
If-else statement looks exactly the same except source timed class. Can be extracted into separated method where `Timed` passed as argument. Or alternatively, having local variable like: 

```
        Timed sourceTimed = methodLevelTimed ! = null ? methodLevelTimed : classLevelTimed;
        builder.description(sourceTimed.description().isEmpty() ? null : sourceTimed.description())
        builder.publishPercentileHistogram(sourceTimed.histogram())
        builder.publishPercentiles(sourceTimed.percentiles().length == 0 ? null : methodLevelTimed.percentiles());
```